File: payment-service/src/main/java/com/hedvig/paymentservice/web/dtos/PaymentMemberDTO.java
package com.hedvig.paymentservice.web.dtos;

import com.hedvig.paymentservice.query.member.entities.Member;
import lombok.Value;

import java.util.Map;
import java.util.UUID;
import java.util.stream.Collectors;

@Value
public class PaymentMemberDTO {

  String id;
  Map<UUID, TransactionDTO> transactions;
  boolean directDebitMandateActive;
  String trustlyAccountNumber;

  public static PaymentMemberDTO fromMember(Member m) {
    return new PaymentMemberDTO(
      m.getId(),
      m.getTransactions().entrySet().stream().collect(
        Collectors.toMap(Map.Entry::getKey, e -> TransactionDTO.fromTransaction(e.getValue()))),
      m.isDirectDebitMandateActive(),
      m.getTrustlyAccountNumber()
    );
  }
}

File: payment-service/src/main/java/com/hedvig/paymentservice/web/dtos/PaymentMemberDTO.kt
package com.hedvig.paymentservice.web.dtos

import com.hedvig.paymentservice.domain.payments.DirectDebitStatus
import com.hedvig.paymentservice.query.member.entities.Member
import java.util.UUID

data class PaymentMemberDTO(
    val id: String,
    val transactions: Map<UUID, TransactionDTO>,
    val directDebitMandateActive: Boolean,
    val trustlyAccountNumber: String? = null
) {

  companion object {
    fun fromMember(member: Member, latestDirectDebitAccountOrderDTO: DirectDebitAccountOrderDTO?) = PaymentMemberDTO(
      id = member.getId(),
      transactions = member.transactions.entries.map { entry -> entry.key to TransactionDTO.fromTransaction(entry.value) }.toMap(),
      directDebitMandateActive = latestDirectDebitAccountOrderDTO?.directDebitStatus == DirectDebitStatus.CONNECTED,
      trustlyAccountNumber = latestDirectDebitAccountOrderDTO?.trustlyAccountId
    )
  }
}
